### PR TITLE
Support LLVM trunk

### PR DESCRIPTION
**Pull request by yzhliu**
LLVM trunk (10.0) requires c++14
(I don't find the exact "announcement", but here is a related one, https://lists.llvm.org/pipermail/llvm-dev/2018-October/127045.html )

@zhiics @icemelon9 @jroesch @tqchen Please review.

### CODE REVIEW

**Comment by icemelon9** [comment]
Should we use c++14 by default when available?

**Comment by MarisaKirisame** [comment]
personally I dont see why not. python2 is deprecated and C++14 is out 5 years ago. giving how fast deep learning is moving I see no incentive to restrict ourselves to 11.

**Comment by zhiics** [comment]
I think C++14 is the default since clang-6.0. But will our submodules cause problem?

**Comment by yzhliu** [comment]
One concern is that, MXNet is using c++11, and for the recently merged feature (use tvm to write mxnet op), we need to compile TVM in MXNet, it will cause problem.

**Comment by yzhliu** [comment]
And some internal systems are still using gcc 4.8/4.9 as I know.

File: CMakeLists.txt [hunk 1]
```diff
@@ -90,15 +90,13 @@ if(MSVC)
     endforeach(flag_var)
   endif()
 else(MSVC)
-  include(CheckCXXCompilerFlag)
-  check_cxx_compiler_flag("-std=c++11"    SUPPORT_CXX11)
   if ("${CMAKE_BUILD_TYPE}" STREQUAL "Debug")
     message("Build in Debug mode")
     set(CMAKE_C_FLAGS "-O0 -g -Wall -fPIC ${CMAKE_C_FLAGS} -rdynamic")
-    set(CMAKE_CXX_FLAGS "-O0 -g -Wall -fPIC -std=c++11 ${CMAKE_CXX_FLAGS} -rdynamic")
```
Comment (zhiics) [inline CMakeLists.txt]: Should we revert this then?

Comment (yzhliu) [reply CMakeLists.txt]: why? I move it to the `if(NOT MSVC) ...` part below.

Comment (zhiics) [reply CMakeLists.txt]: I see. That should be fine. Sorry about it.

**Comment by yzhliu** [comment]
Merge it for now as I need it for feature deliver. Thanks @MarisaKirisame @icemelon9 @zhiics I can open an RFC to discuss the timeline to move to c++14/17

**Pull request merged**